src/hooks/useMutex.js

```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+
+function useMutex() {
+  const queue = React.useRef([]);
+  const lock = React.useRef(false);
+
+  const wait = React.useCallback(() => {
+    const next = queue.current.shift();
+    if (!next) {
+      return;
+    }
+    lock.current = true;
+    return Promise.resolve(next());
+  }, []);
+
+  const signal = React.useCallback(() => {
+    lock.current = false;
+    return wait();
+  }, [wait]);
+
+  const synchronize = React.useCallback(
+    runnable => {
```

Comment:
`synchronize` is the operation that adds a task to the queue. It will execute the task if the lock is not acquired.